src/controllers/HierarchicalFacetSearchController.ts

```diff
@@ -0,0 +1,59 @@
+import { IDynamicHierarchicalFacet } from '../ui/DynamicHierarchicalFacet/IDynamicHierarchicalFacet';
+import { FacetSearchType, IFacetSearchRequest } from '../rest/Facet/FacetSearchRequest';
+import { IFacetSearchResponse } from '../rest/Facet/FacetSearchResponse';
+import { QueryUtils, DateUtils } from '../Core';
+import { FileTypes } from '../ui/Misc/FileTypes';
+
+export class HierarchicalFacetSearchController {
+  constructor(private facet: IDynamicHierarchicalFacet) {}
+
+  private getMonthsValueCaptions() {
+    const monthsValueCaptions: Record<string, string> = {};
+    for (let month = 1; month <= 12; month++) {
+      const key = `0${month}`.substr(-2);
+      monthsValueCaptions[key] = DateUtils.monthToString(month - 1);
+    }
+
+    return monthsValueCaptions;
+  }
+
+  private addTypesCaptionsIfNecessary(): Record<string, string> {
+    const field = this.facet.options.field.toLowerCase();
+    const isFileType = QueryUtils.isStratusAgnosticField(field, '@filetype');
+    const isObjectType = QueryUtils.isStratusAgnosticField(field, '@objecttype');
+    const isMonth = QueryUtils.isStratusAgnosticField(field, '@month');
+
+    if (isFileType || isObjectType) {
+      return FileTypes.getFileTypeCaptions();
+    }
+
+    if (isMonth) {
+      return this.getMonthsValueCaptions();
+    }
+
+    return {};
+  }
+
+  private get captions(): Record<string, string> {
+    return {
+      ...this.addTypesCaptionsIfNecessary(),
+      ...this.facet.options.valueCaption
+    };
+  }
+
+  public search(terms?: string): Promise<IFacetSearchResponse> {
+    const request: IFacetSearchRequest = {
+      field: this.facet.fieldName,
+      type: FacetSearchType.hierarchical,
+      numberOfValues: this.facet.options.numberOfValues,
+      ignorePaths: [this.facet.values.selectedPath],
```

Comment:
`ignorePaths` doesn't seem to work yet. Adding this line here would (supposedly) remove the selected value and its ancestors. Its cousins would still be searchable. Would we rather eliminate every ancestor & cousin?

Reply:
You would want to put every current visible paths in the facet. The goal of ignorePaths/values is to not have the same results in the search as the ones that are already there, that's why in FacetSearchController it's `this.facet.values.allValues`

Reply:
Oh woops, you're right. I used to have it set to `allValues`, but I forgot the purpose of `ignorePaths`. I'm going to fix it.